mypy/newsemanal/semanal.py

```diff
@@ -3862,46 +3862,70 @@ def check_for_obsolete_short_name(self, name: str, ctx: Context) -> None:
-                        if namespace == self.cur_mod_id:
-                            names = self.globals
-                        n = names.get(parts[i], None)
-                        if (not n
-                                and '__getattr__' in names
-                                and not self.is_incomplete_namespace(namespace)):
-                            fullname = namespace + '.' + '.'.join(parts[i:])
-                            gvar = self.create_getattr_var(names['__getattr__'],
-                                                           parts[i], fullname)
-                            if gvar:
-                                n = SymbolTableNode(GDEF, gvar)
-                    # TODO: What if node is Var or FuncDef?
-                    # Currently, missing these cases results in controversial behavior, when
-                    # lookup_qualified(x.y.z) returns Var(x).
-                    if not n:
-                        if not suppress_errors:
-                            self.name_not_defined(name, ctx, namespace=namespace)
-                        break
-                if n:
-                    if n and n.module_hidden:
+        parts = name.split('.')
+        namespace = self.cur_mod_id
+        sym = self.lookup(parts[0], ctx, suppress_errors=suppress_errors)
+        if sym:
+            for i in range(1, len(parts)):
+                node = sym.node
+                if isinstance(node, TypeInfo):
+                    nextsym = node.get(parts[i])
+                elif isinstance(node, MypyFile):
+                    nextsym = self.get_module_symbol(node, parts[i:])
+                    namespace = node.fullname()
+                elif isinstance(node, PlaceholderNode):
+                    return sym
+                else:
+                    if isinstance(node, Var) and isinstance(node.type, AnyType):
+                        # Allow access through Var with Any type without error.
+                        return self.implicit_symbol(sym, name, parts[i:], node.type)
+                    # Lookup through invalid node, such as variable or function
+                    nextsym = None
+                if not nextsym or nextsym.module_hidden:
+                    if not suppress_errors:
                         self.name_not_defined(name, ctx, namespace=namespace)
-            if n and not n.module_hidden:
-                return n
-            return None
+                    return None
+                sym = nextsym
+        return sym
+
+    def get_module_symbol(self, node: MypyFile, parts: List[str]) -> Optional[SymbolTableNode]:
+        """Look up a symbol from the module symbol table."""
+        # TODO: Use this logic in more places?
+        module = node.fullname()
+        names = node.names
+        # Rebind potential references to old version of current module in
+        # fine-grained incremental mode.
+        if module == self.cur_mod_id:
+            names = self.globals
```

Comment:
Is this actually needed in the new analyzer?

Reply:
I don't think that anything has changed here. During a fine-grained update we have both the old and the new module object alive during semantic analysis.

Reply:
@JukkaL do we have a test for this? I tried to just remove these two lines of code and all tests passed. (See also a TODO in `visit_member_expr()`). Moreover, I removed also removed `rebind_symbol_table_node()` completely and all tests still passed.

Should we just remove it from the new analyzer?